File: todo.py
import os
import sys
from datetime import datetime

# Import click
try:
    import click
except ImportError:
    # Install click package if it doesn't exist
    from io import StringIO

    from pip._internal import main as pip

    # Redirecting Standard output and error to String Streams
    sys.stderr, sys.stdout = StringIO(), StringIO()
    pip(['install', '--user', 'click'])

    # Redirect standard output and error back to console
    sys.stdout = sys.__stdout__
    sys.stderr = sys.__stderr__
    import click

# Function to get help text
def get_help():
    commands = [
        'Usage :-',
        '$ ./todo add "todo item"  # Add a new todo',
        '$ ./todo ls               # Show remaining todos',
        '$ ./todo del NUMBER       # Delete a todo',
        '$ ./todo done NUMBER      # Complete a todo',
        '$ ./todo help             # Show usage',
        '$ ./todo report           # Statistics',
    ]
    print("\n".join(commands))
    sys.exit(0)


def add_todo_item(todo_item):
    with open('todo.txt', 'a+') as f:
        f.write(todo_item + "\n")
        print(f'Added todo: "{todo_item}"')


def delete_todo(todo_number):
    try:
        with open('todo.txt', 'r+') as f:
            data = f.readlines()
        if todo_number == 0:
            raise IndexError
        item = data.pop(todo_number - 1)
        with open('todo.txt', 'w') as f:
            f.writelines(data)
        return item
    except (IndexError, FileNotFoundError):
        return f'Error: todo #{todo_number} does not exist.'


def show_pending_todos():
    try:
        with open('todo.txt') as f:
            data = f.readlines()
    except FileNotFoundError:
        print('There are no pending todos!')
        sys.exit(0)

    if len(data) == 0:
        print('There are no pending todos!')
    data = data[::-1]

    pending_todos = len(data)
    for i in range(pending_todos):
        print(f'[{pending_todos-i}] {data[i].strip()}')


# Mark a todo as Done
def complete_todo(todo_number):
    # Call delete_todo to remove item from todo list
    todo_item = delete_todo(todo_number)
    # Check if there was an error
    if todo_item.startswith('Error'):
        print(todo_item)
        sys.exit(0)

    # Get current date in UTC format
    today_date = datetime.utcnow().strftime('%Y-%m-%d')
    with open('done.txt', 'a+') as f:
        f.write(f'x {today_date} {todo_item}')


# Get Statistics
def get_report():
    try:
        with open('todo.txt', 'r+') as f:
            pending = len(f.readlines())
    except FileNotFoundError:
        pending = 0
    try:
        with open('done.txt', 'r+') as f:
            completed = len(f.readlines())
    except FileNotFoundError:
        completed = 0
    today_date = datetime.utcnow().strftime('%Y-%m-%d')
    print(f'{today_date} Pending : {pending} Completed : {completed}')


@click.command()
@click.argument('arg', nargs=-1)  # Taking all arguments provided in a tuple
def main(arg):
    # show help to user
    if len(arg) == 0 or arg[0] == 'help':
        get_help()

    # Add item to todo
    elif arg[0] == 'add':
        try:
            add_todo_item(arg[1])
        except IndexError:
            print('Error: Missing todo string. Nothing added!')

    # Remove item from todo
    elif arg[0] == 'del':
        try:
            number = int(arg[1])
            todo_item = delete_todo(number)
            if todo_item.startswith('Error'):
                print(todo_item, "Nothing deleted.")
            else:
                print(f'Deleted todo #{number}')
        except ValueError:
            print(f'Error: Please provide a valid number to delete item')
        except IndexError:
            print('Error: Missing NUMBER for deleting todo.')

    # Show pending todos
    elif arg[0] == 'ls':
        show_pending_todos()

    # Mark a todo as done
    elif arg[0] == 'done':
        try:
            number = int(arg[1])
            complete_todo(number)
            print(f'Marked todo #{number} as done.')
        except IndexError:
            print('Error: Missing NUMBER for marking todo as done.')

    # Get Report of Pending and Completed todos
    elif arg[0] == 'report':
        get_report()

    # Giving help when a wrong option is selected
    else:
        print('Please select valid option')
        get_help()


if __name__ == '__main__':
    main()

File: todo.js
const { profileEnd } = require('console');
const fs = require('fs');

// save: save current list
function save(data=[], file) {
  return fs.writeFileSync(file, data)
}

// load: load from JSON
function load(file) {
  const text = fs.readFileSync(file, 'utf-8')
  return text.trim()
}

// help: list options
function help() {
  return process.stdout.write(
    `Usage :-
$ ./todo add "todo item"  # Add a new todo
$ ./todo ls               # Show remaining todos
$ ./todo del NUMBER       # Delete a todo
$ ./todo done NUMBER      # Complete a todo
$ ./todo help             # Show usage
$ ./todo report           # Statistics`
  )
}

// built-in functions
const func = {

  // list: display all entries
  ls: function() {
    try{
      let data = load(file='todo.txt').split('\n')
      if(data.length == 0){
        process.stdout.write("There are no pending todos!")
      }
      for(let i=data.length-1; i>=0; i--){
        process.stdout.write("["+`${i+1}`+"] "+`${data[i]}`+"\n")
      }
    }
    catch(err){
      process.stdout.write("There are no pending todos!")
    }
  },

  // add: add entry
  add: function(...title) {

    if(title === undefined || title==''){
      process.stdout.write(`"Error: Missing todo string. Nothing added!"`)
      return
    }
    // let data = load()
    fs.appendFile('todo.txt', title+"\n", function(err) {
      if (err) throw err;
    });
    process.stdout.write(`Added todo: "${title}"`)
    return
  },

  // remove: remove entry
  del: function(term) {
    if(term == undefined){
      process.stdout.write(`"Error: Missing NUMBER for deleting todo."`)
      return
    }
    try{
      term=term-1
      let data = load(file='todo.txt').split('\n')
      if(term > data.length-1 || term == -1 || data[0] == ''){
        process.stdout.write(`"Error: todo #${term+1} does not exist. Nothing deleted."`)
        return
      }

      data.splice(term,1)
      process.stdout.write(`"Deleted todo #${term+1}"`)
      return save(data.join("\n")+"\n",file="todo.txt")
    }
    catch(err){
      process.stdout.write(`"Error: todo #${term} does not exist. Nothing deleted."`)
    }
  },

  // check: check entry
  done: function(id) {
    if(id == undefined){
      process.stdout.write("Error: Missing NUMBER for marking todo as done.")
      return
    }
    try{
      id = id-1
      let data = load('todo.txt').split('\n')
      if(id+1 > data.length || id+1 == 0 || data[0] == ''){
        process.stdout.write(`"Error: todo #${id+1} does not exist."`)
        return
      }
      var todo=data.splice(id,1)
      process.stdout.write(`"Marked todo #${id+1} as done."`)
      if(data.length == 0){
        save("",file="todo.txt")
      }
      else{
        save(data.join("\n")+"\n",file="todo.txt")
      }
      var date = new Date()
      var done = "x "+date.toISOString().split('T')[0]+" "+todo
      fs.appendFile("done.txt",done+"\n", function(err){
        if (err) throw err;
      });

      return
    }
    catch(err){
      throw err
      process.stdout.write(`"Error: todo #${id+1} does not exist."`)
    }
  },

  report: function(){
    try{
      let pending_data = load('todo.txt').split('\n')
      var pcount = pending_data.length
    }
    catch(err){
      var pcount = 0
    }

    try{
      let complete_data = load('done.txt').split('\n')
      var ccount = complete_data.length
    }
    catch(err){
      var ccount = 0
    }
    var date = new Date()
    process.stdout.write(`${date.toISOString().split('T')[0]} Pending : ${pcount} Completed : ${ccount}\n`)
  }

}

// Run from command-line
if (process.argv[2]) {
  const cli = process.argv[2].split(' ')
  if (func[cli[0]]) {
    func[cli[0]](process.argv[3])
  } else {
    help()
  }
} else {
  help()
}
